### ISSUE TITLE
GUI - Selecting a range of values to re-code e.g with the shift key

### ISSUE THREAD

**Issue by mwelch750** [sdcTools/sdcMicro#117]
I am pretty sure we could select a range of values by holding down the shift key during recoding of categorical variables in earlier version. This seems to not be the case anymore.
I am using the Chrome browser.

I tried to do this with the CaseA.dta data supplied before: I wanted to select 102 to 106 and recode to a single category. I was not able to do so - it selects 102 & 106.

require(sdcMicro)
inputdata <- readMicrodata(path="/tmp/Rtmpg0p3yZ/6d878c23938578084556a973/CaseAlab.dta", type="stata", convertCharToFac=TRUE, drop_all_missings=TRUE)
inputdataB <- inputdata

## Convert a numeric variable to factor with user-specified breaks
inputdata <- varToFactor(obj=inputdata, var=c("AGEYRS"))
## Convert a numeric variable to factor with user-specified breaks
inputdata <- varToFactor(obj=inputdata, var=c("EDYRS"))
## Set up sdcMicro object
sdcObj <- createSdcObj(dat=inputdata,
        keyVars=c("REGION","DIST","GENDER","MARITAL","AGEYRS","EDYRS"),
        numVars=c("INCTOTGROSSHH"),
        weightVar=c("WGTPOP"),
        hhId=c("IDH"),
        strataVar=NULL,
        pramVars=NULL,
        excludeVars=NULL,
        seed=0,
        randomizeRecords=FALSE,
        alpha=c(0))

## Recode variable
sdcObj <- groupAndRename(obj=sdcObj, var="DIST", before=c("101","104"), after=c("101_104"))

**Comment by bernhard-da** [comment]
yes, this was possible, but not useful because it lead to racing conditions. I think the current implementation is MUCH better for two reasons.

1) there is no chance, that a race condition (GUI hangs) can occur
2) in the use case where you want to combine many levels, you are almost as fast, but you can "search" by just starting to type in the select field
3) it allows to fix the problem with the old version (where you could use the shift-key) that after each selection, the selection window jumps to the top. This is nasty in case when you have a lot of levels.

to summarize: the current implementation is less error-prone (race-condition) and in my opinion also easier to explain and most of the time faster than the old one (`selectInput(...,selectize=FALSE)`).

**Comment by mwelch750** [comment]
Agreed in general, but very ineffective and time consuming in cases where lists are very long.the example I used has 154 values. There may be cases where there are many more. Selecting each individually may be torture and prone to error.

I understand the racing problem prohibits the elegant solution. What about allowing a range to be enters manually then? Either using inequality , greater , less than or a more simple allowing the start and end bounds to be entered? Is that hard to do?

Best
Matthew

**Comment by bernhard-da** [comment]
Some points I want to add to this discussion.

First off, I think the "ineffectiveness" comes from the workflow you have which is reading in a numeric variable and just use the automatic to factor conversion. This (possibly) leads to many distinct levels in the resulting factor. I agree that in this case, it is a bit more complicated to click a couple of times on the desired levels that you want to combine than to just select with shift and click. But this is only the case, if the levels you want to combine are directly next to each other. Once they are spread over the list, the "old" approach takes much more time because the input always jumps to the top.

Also, in the example that you described, it would probably be much easier to use the manual recoding possibility in the first step. There you can create (among other possibilibies) a factor from a numeric variable by entering the desired groups in the following way: 

 `-Inf,10,20,50,Inf`

This would get result in a factor with the following levels:

```
(-Inf,10]   (10,20]   (20,50] (50, Inf] 
```

which would be way easier to recode/rename than doing the exact thing on the original numeric vector which might have many distinct values when choosing the automatic recode function.

Also, It is not trivial to change the behaviour completely because the underlying workhorse-function `groupAndRename()` deals with factors and its levels only. So the factor-levels might not be able to be converted back to numbers in which case the suggested enhancement using logical conditions would not work.

Also, in the TORS we have that data manipulation can be done outside of the GUI in whatever software the analyst is comfortable with. I think the functionality that is already included in the GUI is quite a lot and if additional steps would be required, it might be easier for analysts to do them in (any) software but not in the GUI.

**Comment by mwelch750** [comment]
Dear Bernhardt,

I am perfectly comfortable myself doing most of not all functions outside the GUI, but it is my task to think from the perspective of our clients and the level of their skills. To the extent it is possible to make it easier for them I would like to do so.  Just to be clear this is not my workflow, but one which we have discussed as the easiest to implement. If there is a better way we are always happy to hear about that.

I am not averse to giving proper instructions to users as to how things are best approached. So on the screen where the numeric is converted to factor we should be much clearer about why you might favor grouping at that stage already over automatic conversion and then doing later. One question to answer is: if we care about measuring risk of the original ungrouped variable as a base then how do we do this if we recode before setting up the problem or even if we do this outside the application. How do we report on or more precisely how does the statistical office report on risk of the raw data over the anonymize data  effectively within the confines of this tool?

Yes, the TORs mention doing manipulations outside the GUI for data manipulation outside the scope of the application. I did not envisage recoding falling into that category as it is a valid anonymization method.

I don't think this is a big problem , but it would be kind to think how we help our users understand what needs to be done in a self explanatory way. I think you have done a fantastic job and you have certainly been very patient with us. We are very excited by the progress and are looking forward to making it a polished software product that agencies can use to unlock access to data which remains locked up due to skill barriers preventing adequate anonymization to allow release.

Best
Matthew

**Comment by mwelch750** [comment]
Moved response to #117 to email
